aten/src/ATen/native/native_functions.yaml

```diff
@@ -6389,19 +6389,38 @@
     CPU: sort_out_cpu
     CUDA: legacy::cuda::_th_sort_out
 
+- func: sort.values_stable(Tensor self, *, bool? stable, int dim=-1, bool descending=False, Tensor(a!) values, Tensor(b!) indices) -> (Tensor(a!) values, Tensor(b!) indices)
+  use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
+  dispatch:
+    CPU: sort_out_cpu_stable
+    CUDA: legacy::cuda::_th_sort_out_stable
+
 - func: sort(Tensor self, int dim=-1, bool descending=False) -> (Tensor values, Tensor indices)
   variants: method, function
   dispatch:
     CPU: sort_cpu
     CUDA: legacy::cuda::_th_sort
     QuantizedCPU: sort_quantized_cpu
 
+- func: sort.stable(Tensor self, *, bool? stable, int dim=-1, bool descending=False) -> (Tensor values, Tensor indices)
```

Comment:
There's another issue with this PR. In NumPy the argument is called "kind" and accepts a string:

https://numpy.org/doc/stable/reference/generated/numpy.sort.html

Further question, when will stable sort produce a different result than regular sort?

Reply:
Stable will produce different results from qsort, for example, with repeated keys, as it never swaps values with the same key value.
Currently, on the CPU, we use the STL's sort, and it appears to be adaptive, although in most cases it is a qsort. The STLs stable sort also appears to be adaptive, being either an out- or in-place. So, it seems like we have only two kinds of adaptive sorting routines :)

Reply:
I guess I'm curious what the point of implementing stable sort in PyTorch is, then?

Reply:
I guess one of the use case is the `isin` function which need a stable sort on 1d tensor.

Reply:
Can you elaborate with an example of what the stable sort does differently?

Reply:
Found this: https://github.com/pytorch/pytorch/issues/28871

Reply:
Got it! Also looks like this PR should address that issue, too?

I think we should update the docs with an example. This makes sense because sort returns both the sorted array + the original indices of the sorted elements.

Reply:
Example will have to be large enough. For small inputs std::sort is using an insertion sort and is almost always stable.

Reply:
Under GNU C++ (use introsort), any length that larger than 7 should be fine, as long as `log10(length of array) * 2 >= 1.0`, the `std::sort()` will be unstable.

Reply:
I got the following minimal repro for binary tenors:
```
In [19]: x = torch.tensor([0, 1] * 9)

In [20]: x.sort()
Out[20]: 
torch.return_types.sort(
values=tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1]),
indices=tensor([ 2, 16,  4,  6, 14,  8,  0, 10, 12,  9, 17, 15, 13, 11,  7,  5,  3,  1]))

In [21]: x.sort(stable=True)
Out[21]: 
torch.return_types.sort(
values=tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1]),
indices=tensor([ 0,  2,  4,  6,  8, 10, 12, 14, 16,  1,  3,  5,  7,  9, 11, 13, 15, 17]))

```

Reply:
> There's another issue with this PR. In NumPy the argument is called "kind" and accepts a string:
> 
> https://numpy.org/doc/stable/reference/generated/numpy.sort.html
> 

I guess this is the last thing. @mruberry should we use kind + string type for numpy compatibility? It's a bit strange to me to let c++ code passing a string type flag. Maybe we can just keep this boolean flag? Any comments?

Reply:
My 2 cents: we do not offer that many options unlike NumPy, and both options, stable or not, are composite algorithms which have different runtime depending on memory available. I would vote for a stable flag.

Reply:
@nikitaved 
Yep, let's keep this implementation for now.